pkg/quarterdeck/auth.go

```diff
@@ -637,6 +639,126 @@ func (s *Server) Refresh(c *gin.Context) {
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "switch to same org").Inc()
+		return
+	}
+
+	// Verify that the user belongs to the requested organization.
+	var orgUser *models.OrganizationUser
+	if orgUser, err = models.GetOrgUser(c.Request.Context(), claims.Subject, in.OrgID); err != nil {
+		if errors.Is(err, db.ErrNotFound) {
+			sentry.Warn(c).Str("userID", claims.Subject).Str("orgID", in.OrgID.String()).Msg("user attempt to switch into organization they do not belong to")
+			c.JSON(http.StatusForbidden, api.ErrorResponse("invalid credentials"))
+			metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "invalid credentials").Inc()
+			return
+		}
+
+		sentry.Error(c).Err(err).Msg("could not retrieve organization user from database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not process switch request"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "unhandled error").Inc()
+		return
+	}
+
+	// Fetch the user information and create claims now that they are successfully authenticated
+	var user *models.User
+	if user, err = orgUser.User(c.Request.Context(), false); err != nil {
+		sentry.Error(c).Err(err).Msg("could not retrieve user from database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not process switch request"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "unhandled error").Inc()
+		return
+	}
+
+	// Create access and refresh tokens for new organization
+	newClaims := &tokens.Claims{
+		RegisteredClaims: jwt.RegisteredClaims{
+			Subject: user.ID.String(),
+		},
+		Name:    user.Name,
+		Email:   user.Email,
+		Picture: gravatar.New(user.Email, nil),
+		OrgID:   orgUser.OrgID.String(),
+	}
+
+	// Add the user permissions to the claims
+	if claims.Permissions, err = user.Permissions(c.Request.Context(), false); err != nil {
+		sentry.Error(c).Err(err).Msg("could not get user permissions from model")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not create credentials"))
+
+		// increment failure count
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "permissions not found").Inc()
+		return
+	}
+
+	out = &api.LoginReply{
+		LastLogin: user.LastLogin.String,
+	}
+	if out.AccessToken, out.RefreshToken, err = s.tokens.CreateTokenPair(newClaims); err != nil {
+		sentry.Error(c).Err(err).Msg("could not create access and refresh token")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not create credentials"))
+
+		// increment failure count
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "jwt token error").Inc()
+		return
+	}
+
+	// Update the user's last login in a Go routine so it doesn't block
+	s.tasks.QueueContext(sentry.CloneContext(c), tasks.TaskFunc(func(ctx context.Context) error {
+		ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+		defer cancel()
+		return user.UpdateLastLogin(ctx)
```

Comment:
Do we have to update the organization last login as well?

Reply:
This method was modified to also update the last login on the user's "loaded" organization, so we should be covered.

Reply:
Perfect!